File: src/main/java/cards/SpikedShield.java
package cards;

import basemod.abstracts.CustomCard;
import com.megacrit.cardcrawl.actions.common.DamageAction;
import com.megacrit.cardcrawl.actions.common.GainBlockAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import enums.AbstractCardEnum;
import initliziers.ThePainInitializer;

public class SpikedShield extends CustomCard {

    private static final int COST = 1;
    private static final int BLOCK_AMT = 10;
    private static final int UPGRADE_PLUS_BLOCK = 2;
    private static final int SELF_DMG_AMT = 2;
    private static final int UPGRADE_SELF_DMG_ATM = 2;

    public static final String ID = "Spiked_Shield";
    public static final String NAME = "Spiked Shield";
    public static final String DESCRIPTION = "Gain !B! Block. NL Take !M! Damage.";

    public SpikedShield() {
        super(ID, NAME, ThePainInitializer.DEFAULT_CARD_IMAGE_PATH, COST, DESCRIPTION, CardType.SKILL, AbstractCardEnum.THE_PAIN_PURPLE, CardRarity.COMMON, CardTarget.SELF);
        this.block = this.baseBlock = BLOCK_AMT;
        this.magicNumber = this.baseMagicNumber = SELF_DMG_AMT;
    }

    @Override
    public void upgrade() {
        if(!upgraded) {
            upgradeName();
            upgradeBlock(UPGRADE_PLUS_BLOCK);
            upgradeMagicNumber(UPGRADE_SELF_DMG_ATM);
        }
    }

    @Override
    public AbstractCard makeCopy() {
        return new SpikedShield();
    }

    @Override
    public void use(AbstractPlayer abstractPlayer, AbstractMonster abstractMonster) {
        AbstractDungeon.actionManager.addToBottom(new DamageAction(abstractPlayer, new DamageInfo(abstractPlayer, this.magicNumber, DamageInfo.DamageType.HP_LOSS)));
        AbstractDungeon.actionManager.addToBottom(new GainBlockAction(abstractPlayer, abstractPlayer, this.block));
    }
}

File: src/main/kotlin/kobting/thepain/cards/SpikedShield.kt
package kobting.thepain.cards

import com.megacrit.cardcrawl.actions.common.DamageAction
import com.megacrit.cardcrawl.actions.common.GainBlockAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.DamageInfo
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import kobting.thepain.patches.AbstractCardEnum

class SpikedShield : PainCustomCard(ID, COST, AbstractCard.CardType.SKILL, AbstractCardEnum.THE_PAIN_PURPLE, AbstractCard.CardRarity.COMMON, AbstractCard.CardTarget.SELF) {
    init {
        this.baseBlock = BLOCK_AMT
        this.block = this.baseBlock
        this.baseMagicNumber = SELF_DMG_AMT
        this.magicNumber = this.baseMagicNumber
    }

    override fun upgrade() {
        if (!upgraded) {
            upgradeName()
            upgradeBlock(UPGRADE_PLUS_BLOCK)
            upgradeMagicNumber(UPGRADE_SELF_DMG_ATM)
        }
    }

    override fun makeCopy(): AbstractCard {
        return SpikedShield()
    }

    override fun use(abstractPlayer: AbstractPlayer?, abstractMonster: AbstractMonster?) {
        AbstractDungeon.actionManager.addToBottom(DamageAction(abstractPlayer, DamageInfo(abstractPlayer, this.magicNumber, DamageInfo.DamageType.HP_LOSS)))
        AbstractDungeon.actionManager.addToBottom(GainBlockAction(abstractPlayer, abstractPlayer, this.block))
    }

    companion object {

        val ID = "thepain:SpikedShield"

        private val COST = 1
        private val BLOCK_AMT = 10
        private val UPGRADE_PLUS_BLOCK = 2
        private val SELF_DMG_AMT = 2
        private val UPGRADE_SELF_DMG_ATM = 2
    }
}
